PC/winreg.c

```diff
@@ -143,8 +143,10 @@ PyHKEY_deallocFunc(PyObject *ob)
        check to fail!
     */
     PyHKEYObject *obkey = (PyHKEYObject *)ob;
+#ifndef MS_WINDOWS_NON_DESKTOP
```

Comment:
The `winreg` module should not be built for non-desktop targets, *except* for games. For whatever reason, "winreg.h" in version 10.0.19041 of the SDK adds significant support for the games API partition. All of the following functions are available:

```text
RegCloseKey
RegCreateKeyExW
RegCreateKeyW
RegDeleteKeyExW
RegDeleteKeyW
RegDeleteValueW
RegEnumKeyExW
RegEnumValueW
RegOpenKeyExW
RegQueryInfoKeyW
RegQueryValueExW
RegSetValueExW
```

The following registry functions are only defined in the desktop and system partitions:

```text
RegConnectRegistryW
RegDisableReflectionKey
RegEnableReflectionKey
RegFlushKey
RegLoadKeyW
RegQueryReflectionKey
RegQueryValueW
RegSaveKeyW
RegSetValueW
```

Tests will have to be skipped or modified if they directly test or depend on a `winreg` function that's not available on Xbox.

Reply:
1) what exception should the non existing throw? E.g. `winreg_DisableReflectionKey_impl` uses NotImplementedError. However this would be a new exception for many of these APIs
2) this will require changes at least to `_bootstrap_external.py`
3) at least right now I am only running an embedded python interpreter in our game engine on the xbox with frozen modules. So at this point I do not run any tests, even though it would be obviously great to run at least part of the test suite on there.

Reply:
Off the top of my head:
* `RegConnectRegistryW` - raise NotImplementedError
* `RegDisableReflectionKey` - raise OSError
* `RegEnableReflectionKey` - raise OSError
* `RegFlushKey` - no-op
* `RegLoadKeyW` - raise NotImplementedError
* `RegQueryReflectionKey` - raise OSError
* `RegQueryValueW` - reimplement using `RegQueryValueExW`
* `RegSaveKeyW` - raise NotImplementedError
* `RegSetValueW` - reimplement using `RegSetValueW`

The two reimplementations can be unconditional - there's no reason for us to be using the old APIs, as far as I know.

The ones that should raise OSError are likely continuable. They can use `ERROR_NOT_SUPPORTED` to communicate the reason, and code that expected OS-type failures will handle and continue.

Flushing is not recommended anyway, but shouldn't affect anything. We can just let it do nothing.

The rest seem likely to be non-continuable. If they fail, the code is not going to be able to do what they wanted, and there's nothing they can do about it other than change their code. `NotImplementedError` is the quickest way to make that obvious.

Reply:
You could simply skip defining these functions:

```text
winreg.ConnectRegistry
winreg.DisableReflectionKey
winreg.EnableReflectionKey
winreg.FlushKey
winreg.LoadKey
winreg.QueryReflectionKey
winreg.SaveKey
```

and the exception a script will get is `AttributeError`.

* Connecting to a remote registry, and loading and saving keys are all administrative tasks. They aren't relevant on Xbox.
* Windows 7+ uses key sharing instead of [reflection](https://learn.microsoft.com/en-us/windows/win32/winprog64/registry-reflection), so enabling, disabling, and querying key reflection has no practical value.
* Flushing a key is hardly ever done since the configuration manager flushes keys periodically anyway. It may not even be relevant on Xbox depending on how the registry is implemented.

`winreg.QueryValue()` and `winreg.SetValue()` are the old Win16 query/set value functions that operate on a key's default value. Ideally they shouldn't be used. Unfortunately we haven't highlighted this in the docs. We could implement them in a way that's compatible with Xbox, using WinAPI `RegOpenKeyExW()` (for `sub_key`), `RegQueryValueExW()`, and `RegSetValueExW()`.